File: first-and-last-position-of-element.py
# Time complexity: O(logn+logm)
# Space complexity: O(1)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarysearchfirst(nums, target, low, high):
            while low <= high:
                mid = low + (high-low)//2

                if nums[mid] == target:
                    if (mid==0 or nums[mid] != nums[mid-1]):
                        return mid
                    else:
                        high = mid-1
                elif nums[mid] > target:
                    high = mid -1
                else:
                    low = mid +1
            return -1
        
        def binarysearchlast(nums, target, low, high):
            while low <= high:
                mid = low + (high-low)//2

                if nums[mid] == target:
                    if (mid == len(nums)-1 or nums[mid] != nums[mid+1]):
                        return mid
                    else:
                        low = mid+1
                elif nums[mid] > target:
                    high = mid -1
                else:
                    low = mid +1
            return -1

        low = 0
        high = len(nums)-1
        first = binarysearchfirst(nums, target, low, high)
        if first == -1:
            return [-1,-1]
        last = binarysearchlast(nums, target, first, high)
        return [first, last]

File: FirstAndLastPositionOfElement.java
class FirstAndLastPositionOfElement {

    int findMinMid(int[] nums, int target){
        int low = 0, high = nums.length - 1;
        while(high >= low){
            int mid = (low + high)/2;
            if(nums[mid] == target) {
                if(mid==low || nums[mid]!=nums[mid-1]){
                    return mid;
                }
                high = mid - 1;
            }
            if(target < nums[mid]){
                high = mid - 1;
            }else if (target > nums[mid]){
                low = mid + 1;
            }
        }
        return -1;
    }

    int findMaxMid(int[] nums, int target){
        int low = 0, high = nums.length - 1;
        while(high >= low){
            int mid = (low + high)/2;
            if(nums[mid] == target) {
                if(mid==high || nums[mid]!=nums[mid+1]){
                    return mid;
                }
                low = mid + 1;
            }
            if(target < nums[mid]){
                high = mid - 1;
            }else if (target > nums[mid]){
                low = mid + 1;
            }
        }
        return -1;
    }

    public int[] searchRange(int[] nums, int target) {
        int minMid = findMinMid(nums, target);
        int maxMid = findMaxMid(nums, target);

        return new int[]{minMid, maxMid};
    }
}
